src/query_sink.rs

```diff
@@ -0,0 +1,129 @@
+use winapi::{
+    um::wbemcli::{
+        IWbemClassObject,
+        WBEM_NO_ERROR,
+    },
+    shared::{
+        ntdef::HRESULT,
+        wtypes::BSTR,
+    },
+    ctypes::{
+        c_long,
+    },
+};
+use winapi::um::wbemcli::{IWbemObjectSink, IWbemObjectSinkVtbl};
+use com_impl::{ComImpl, VTable, Refcount};
+use log::{trace, warn};
+use std::ptr::NonNull;
+use wio::com::ComPtr;
+use async_channel::Sender;
+use crate::result_enumerator::IWbemClassWrapper;
+
+/// Implementation for IWbemObjectSink.
+/// This [Sink] receives asynchronously the result of the query,
+/// through Indicate calls. When finished,the SetStatus method
+/// is called.
+/// [Sink]: https://en.wikipedia.org/wiki/Sink_(computing)
+/// # https://docs.microsoft.com/fr-fr/windows/win32/wmisdk/example--getting-wmi-data-from-the-local-computer-asynchronously
+#[repr(C)]
+#[derive(ComImpl)]
+#[interfaces(IWbemObjectSink)]
+pub struct QuerySink {
+    vtbl: VTable<IWbemObjectSinkVtbl>,
+    refcount: Refcount,
+    sender: Sender<Vec<IWbemClassWrapper>>,
+}
+
+impl QuerySink {
+    pub fn new(sender: Sender<Vec<IWbemClassWrapper>>) -> ComPtr<IWbemObjectSink> {
+        let ptr = QuerySink::create_raw(sender);
+        let ptr = ptr as *mut IWbemObjectSink;
+        unsafe { ComPtr::from_raw(ptr) }
+    }
+}
+
+#[com_impl::com_impl]
+unsafe impl IWbemObjectSink for QuerySink {
+    pub unsafe fn indicate(
+        &self,
+        lObjectCount: c_long,
+        apObjArray: *mut *mut IWbemClassObject
+    ) -> HRESULT {
+        trace!("Indicate call with {} objects", lObjectCount);
+        // apObjArray The array memory itself is read-only, and is owned by the caller of the method.
+        // Call AddRef on each element of apObjArray to borrow.
+        // https://docs.microsoft.com/en-us/windows/win32/api/wbemcli/nf-wbemcli-iwbemobjectsink-indicate
+
+        // TODO: Document when ObjectCount is <=0
+        if lObjectCount <= 0 {
+            return WBEM_NO_ERROR as i32;
+        }
+
+        let lObjectCount = lObjectCount as usize;
+        let tx = self.sender.clone();
+        let mut result = Vec::<IWbemClassWrapper>::with_capacity(lObjectCount);
+
+        unsafe {
+            // TODO: check if pointers are non null
+            // Iterate over input array to extract ClassObjects
+            for i in 0..lObjectCount {
+                let p_el = *apObjArray.offset(i as isize);
+                let wbemClassObject = IWbemClassWrapper::new(NonNull::new(p_el));
+                // call AddRef because object will be held after the end of Indicate
+                wbemClassObject.add_ref();
+                result.push(wbemClassObject);
+            }
+        }
+
+        // send the result to the receiver
+        if let Err(e) = tx.try_send(result) {
+            warn!("Error while sending result: {}", e);
```

Comment:
If sending failed, we should report back to WMI (you can add a `todo` for now, but we shouldn't forget!)

Reply:
I don't know if we need to notify WMI in this case, as it's more a problem with our mpsc channel..

However, if a pointer in apObjArray is null, I think we need to notify WMI. I don't know yet which HRESULT to use.
I added a `todo` anyway